### PR TITLE
Thread support

### PR DESCRIPTION
**Pull request by tobbe-j**
Comments displayed in threads, with only the first comment visible on load.

Threads can be opened one at a time, with a submit-field at the bottom of every thread.

For all your own comments, a delete button is visible allowing you to delete the comment,

Comments from the original-poster have an "op" label after their name.

The tag element button has been moved to beside the comment button, under the submit-fields. Elements can be tagged on a thread basis, so you can tag a comment in one thread, open another one and tag another lement there.

TODO: 
 -~~Backend support for dev to delete any comment~~
 -~~Modal before deleting comment~~
 -Support for dev label
 -Support for mulitple labels
 -~~Overall styling improvements, transitions ...~~
 -~~Needs to be integrated with #35~~

### CODE REVIEW

File: client/src/components/comment-label/comment-label.scss [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  color: $palette-white;
+  @include shadow-outline(1px);
+}
+
+.dev {
+  background-color: green;
+}
+
+.op {
+  background-color: dodgerblue;
```
Comment (HenrikHelenius) [inline client/src/components/comment-label/comment-label.scss]: no hardcoded values, move to colors file

File: client/src/components/comment-label/comment-label.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import classNames from 'classnames/bind'
+
+
+// Styles
+import submitFieldStyles from './comment-label.scss'
+
+const css = classNames.bind(submitFieldStyles)
+
+const CommentLabel = ({ posterRole }) => (
+  <div className={css('comment-label', posterRole)}>
```
Comment (HenrikHelenius) [inline client/src/components/comment-label/comment-label.js]: rename to role-label

File: client/src/components/comment-panel/comment-panel.js [hunk 3]
```diff
@@ -80,25 +85,56 @@ class CommentPanel extends React.Component {
...
+  async deleteComment(comment) {
+    if (Object.keys(this.props.users)[0] === comment.userId) {
+      await apiCall('DELETE', '/comments', { commentId: comment.id })
+      await this.fetchComments()
+    }
   }
 
   scrollToBottom() {
-    const el = shadowDocument().getElementById('comment-container')
+    const el = shadowDocument().getElementById('thread--container')
```
Comment (HenrikHelenius) [inline client/src/components/comment-panel/comment-panel.js]: don't use --, confusing for BEM

File: client/src/components/comment-panel/comment-panel.js [hunk 4]
```diff
@@ -80,25 +85,56 @@ class CommentPanel extends React.Component {
...
+    const threadArray = groupByThread(Object.values(this.props.comments))
+    const sortbyTime = R.sortBy(([id, comments]) => R.reduce(
+      R.minBy(comment => comment.time),
+      { time: '9999-99-99 99:99:99' },
+      comments
+    ).time)
+    const sortedThreads = sortbyTime(R.toPairs(threadArray))
     return (
-      <div className={css('comment-container')} id="comment-container">
+      <div className={css('thread-container')} id="thread--container">
```
Comment (HenrikHelenius) [inline client/src/components/comment-panel/comment-panel.js]: see -- ^

File: client/src/components/comment/comment.js [hunk 5]
```diff
@@ -1,13 +1,13 @@
 import React from 'react'
-// Helpers
+import Moment from 'react-moment'
```
Comment (HenrikHelenius) [inline client/src/components/comment/comment.js:3]: move moment under // helpers

File: client/src/components/comment/comment.js [hunk 6]
```diff
@@ -58,6 +61,10 @@ const Comment = ({ id, comment, role }) => (
       </div>
     </div>
     <Reactions reactions={comment.reactions} commentId={id} />
+    <div className={css('expand-button-container')}>
+      <button className={css('delete-button', { hidden: !canDelete })} type="button" onClick={() => deleteComment(comment)}>Delete</button>
```
Comment (HenrikHelenius) [inline client/src/components/comment/comment.js]: break line, too long?

File: client/src/components/comment/comment.scss [hunk 7]
```diff
@@ -52,3 +52,21 @@
...
 }
+
+.expand-button-container {
+  display: flex;
+  justify-content: flex-end;
+
+  button {
+    @include button();
+    background: white;
+    @include shadow-outline(1px);
```
Comment (HenrikHelenius) [inline client/src/components/comment/comment.scss]: move above bg

File: client/src/components/comment/comment.scss [hunk 8]
```diff
@@ -52,3 +52,21 @@
...
+    position: relative;
```
Comment (HenrikHelenius) [inline client/src/components/comment/comment.scss]: order:
includes
position
top
margin
bg 

others

File: client/src/components/tag-element-button/tag-element-button.js [hunk 9]
```diff
@@ -1,6 +1,11 @@
 import React from 'react'
+
```
Comment (HenrikHelenius) [inline client/src/components/tag-element-button/tag-element-button.js]: remove empty line

File: client/src/components/tag-element-button/tag-element-button.scss [hunk 10]
```diff
@@ -4,21 +4,48 @@
...
-  bottom: 12%;
-  left: 3%;
-  $dimension: 30px;
-  height: $dimension;
-  padding: $baseline $spacer;
-  box-shadow: 0 3px 5px $palette-light;
-  color: $palette-lighter;
-  background: $palette-accent;
+  @include button(100%);
+  cursor: pointer;
```
Comment (HenrikHelenius) [inline client/src/components/tag-element-button/tag-element-button.scss]: order: 
include
margin
padding
box-s
bg
cursor

File: client/src/components/tag-element-button/tag-element-button.scss [hunk 11]
```diff
@@ -4,21 +4,48 @@
...
+  background: $palette-white;
   &.active {
-    background: $palette-dark;
-    &, & * {
-      visibility: visible;
+    visibility: visible;
+    @include button(2px);
+    color: $palette-dark;
+    background: $palette-accent;
+    box-shadow: 0 3px 5px $palette-light;
```
Comment (HenrikHelenius) [inline client/src/components/tag-element-button/tag-element-button.scss]: box shadow above color
visibility after include

File: client/src/components/tag-element-button/tag-element-button.scss [hunk 12]
```diff
@@ -4,21 +4,48 @@
...
+
+  }
+  &.selected {
+    background: $palette-medium;
+  }
+  i {
```
Comment (HenrikHelenius) [inline client/src/components/tag-element-button/tag-element-button.scss]: can these be extended? copied from somewhere? @tobbe-j

**Review by HenrikHelenius — DISMISSED** [review]

File: client/src/components/comment-label/comment-label.js (hunk 2 above)
Comment (tobbe-j) [reply client/src/components/comment-label/comment-label.js]: Whole component or only css class?

File: client/src/components/tag-element-button/tag-element-button.js [hunk 13]
```diff
@@ -33,15 +37,15 @@ class TagElementButton extends React.Component {
...
 
   render() {
-    const { active } = this.props
+    const { active, selected } = this.props
 
     return (
       <button
         type="button"
-        className={css('button', { active })}
+        className={css('button', { active }, { selected })}
```
Comment (bqqbarbhg) [inline client/src/components/tag-element-button/tag-element-button.js]: I think these can be in the same object

Comment (bqqbarbhg) [reply client/src/components/tag-element-button/tag-element-button.js]: As in `{ active, selected }`

File: client/src/components/thread/thread.js [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+          inputRef={this.replyField}
+          toggleTagElementState={this.props.toggleTagElementState}
+        />
+      </>
+    )
+  }
+
+  canDelete(commentUserId) {
+    const [currentUser] = Object.keys(this.props.users)
+    return currentUser === commentUserId || this.props.role === 'dev'
```
Comment (bqqbarbhg) [inline client/src/components/thread/thread.js]: This works reliably only with one user.
Use `users.hasOwnProperty(commentUserId)`

**Comment by bqqbarbhg** [comment]
Should the "Expand" button show how many comments are hidden in the thread?

File: client/src/components/comment-label/comment-label.js (hunk 2 above)
Comment (tobbe-j) [reply client/src/components/comment-label/comment-label.js]: Will fix in thread-improvements branch

**Review by quvide — APPROVED** [review]
ok

**Pull request merged**